### ISSUE TITLE
Vertical adjust does not affect maximum horizontal interrupt line

### ISSUE THREAD

**Issue by grauw** [openMSX/openMSX#1106]
In openMSX changing the vertical adjust affects the maximum screen split line, but on a real MSX it does not affect it.

This hung my game project’s code when I changed the adjust to a non-zero value, and also breaks VGMPlay on an MSX2 @ 50 Hz with a vertical adjust > 5.

Test program, should hang at 234:

    10 VDP(10)=0     ' +2 for 50 Hz, +128 for 212 lines
    20 VDP(19)=&H70  ' changing this should have no effect
    30 FOR I=255 to 0 STEP -1
    40 PRINT I
    50 VDP(20)=I
    60 VDP(0)=VDP(0) OR 16
    70 I$=INPUT$(1)
    80 VDP(0)=VDP(0) AND 239
    90 NEXT

Verified on an MSX turboR GT.

Context: https://www.msx.org/forum/msx-talk/development/grauw%E2%80%99s-rpg-in-development?page=14#comment-338647

**Comment by grauw** [comment]
Given that the maximum interrupt line doesn’t change with the vertical adjustment, I thought maybe IL wasn’t affected by vertical adjust, but it is and openMSX emulates that correctly. But I did notice on my real MSX that the top lines repeat when the adjust is 7 while overscanning (except for the last), openMSX doesn’t do this.

I think this could be explained by the VDP resetting its internal line counter up to seven lines later than usual (so the first few get an old value), rather than immediately after top erase with an offset. Which would explain the maximum interrupt line not changing.

In other words, vertical adjust changes the line on which the line counter is reset, rather than the value the line counter is reset to.

Here’s a test ROM; change the vertical adjustment with u and j. When adjusting to the most downwards position, observe the difference for the topmost displayed lines between openMSX and a real MSX2+.
[tiletile.rom-adjust.zip](https://github.com/openMSX/openMSX/files/1985579/tiletile.rom-adjust.zip)

**Comment by MBilderbeek** [comment]
I wonder if this could be related to #471 ...

**Comment by grauw** [comment]
> I wonder if this could be related to #471 ...

I think it is. That report says the issue presents itself depending on the vertical adjust value, which is exactly what this bug is about.

**Comment by MBilderbeek** [comment]
So, any idea what should be changed in the code to test your theory? :) @m9710797 perhaps? Or @mthuurne ?

**Comment by m9710797** [comment]
I completely forgot about this bug. Sorry about that.

How do I run the test ROM from the 2nd post? I tried several romtypes, but none seem to work (all just boot to BASIC).

But the BASIC test prog in the 1st post is also nice. I changed the values on lines 10 and 20 as described in the comments. Here are some results:
* Real TurboR-GT machine:
VDP(0)=0  -> hang at 234
VDP(0)=2  -> hang at 255  (thus immediately hangs)
VDP(0)=128  -> hang at 244
VDP(0)=130  -> hang at 255
* openMSX revision 7b4d9aef
VDP(0)=0  -> hang at 234 + V
VDP(0)=2  -> hang at min(255, 258 + V)   (e.g. for V=-8 hang at 250)
VDP(0)=128  -> hang at 244 + V
VDP(0)=130  -> hang at 255
with V=-8..+7

So I can confirm what grauw said:
* For vertical set-adjust=0, openMSX behaves the same as the real machine.
* On real machine vertical set adjust has no influence. But it wrongly does have in openMSX.

The only new info (or clarification) in this post is that the line on which the hang triggers does depend on the value of VDP(0).

In the next days (next weekend?) I'll try to fix this in openMSX.

**Comment by MBilderbeek** [comment]
@m9710797 Unfortunately, that commit doesn't fix #471 ....  :(

**Comment by grauw** [comment]
> How do I run the test ROM from the 2nd post? I tried several romtypes, but none seem to work (all just boot to BASIC).

-romtype KonamiSCC

Requires a V9958, probably is the reason it won’t start for you.

You may still want to check out the top lines repeating in the overscan area for positive adjust values.

**Comment by MBilderbeek** [comment]
@grauw  @m9710797  What is the status of this ticket?

**Comment by MBilderbeek** [comment]
When I run the test ROM and put the adjust to the most downward position on a turboR, I see this:
![test](https://user-images.githubusercontent.com/12675978/81729467-4c05c400-948c-11ea-9719-9e04b0ebe40a.png)
I understand that this is not what happens on real turboR? I don't have a means to easily test ROM software on my turboR...

**Comment by grauw** [comment]
Iirc what happened on a real turboR is that instead of those purple lines (which is what is in VRAM on the lines above) it repeats the first couple of lines where the image starts. This suggests that the VDP resets the vertical line counter twice when you adjust downward.

**Comment by grauw** [comment]
I moved that to a separate issue #1257. This problem is fixed.
